File: src/scratch/ide/popup/ScratchListElementRenderer.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scratch.ide.popup;

import com.intellij.openapi.ui.popup.ListItemDescriptor;
import com.intellij.openapi.ui.popup.ListPopupStep;
import com.intellij.openapi.ui.popup.MnemonicNavigationFilter;
import com.intellij.openapi.util.IconLoader;
import com.intellij.ui.ColorUtil;
import com.intellij.ui.popup.list.GroupedItemsListRenderer;
import com.intellij.util.ui.UIUtil;

import javax.swing.*;

class ScratchListElementRenderer extends GroupedItemsListRenderer {
	public static final Icon NextStep = IconLoader.findIcon("/icons/ide/nextStep.png");
	public static final Icon NextStepGrayed = IconLoader.getIcon("/icons/ide/nextStepGrayed.png"); // 12x12
	public static final Icon NextStepInverted = IconLoader.getIcon("/icons/ide/nextStepInverted.png"); // 12x12

	private final ScratchListPopup myPopup;

	public ScratchListElementRenderer(final ScratchListPopup aPopup) {
		super(new ListItemDescriptor() {
			@Override
			public String getTextFor(Object value) {
				return aPopup.getListStep().getTextFor(value);
			}

			@Override
			public String getTooltipFor(Object value) {
				return null;
			}

			@Override
			public Icon getIconFor(Object value) {
				return aPopup.getListStep().getIconFor(value);
			}

			@Override
			public boolean hasSeparatorAboveOf(Object value) {
				return aPopup.getListModel().isSeparatorAboveOf(value);
			}

			@Override
			public String getCaptionAboveOf(Object value) {
				return aPopup.getListModel().getCaptionAboveOf(value);
			}
		});
		myPopup = aPopup;
	}

	@Override
	protected void customizeComponent(JList list, Object value, boolean isSelected) {
		ListPopupStep<Object> step = myPopup.getListStep();
		boolean isSelectable = step.isSelectable(value);
		myTextLabel.setEnabled(isSelectable);

		if (step.isMnemonicsNavigationEnabled()) {
			MnemonicNavigationFilter<Object> navigationFilter = step.getMnemonicNavigationFilter();
			final int pos = (navigationFilter == null ? -1 : navigationFilter.getMnemonicPos(value));
			if (pos != -1) {
				String text = myTextLabel.getText();
				text = text.substring(0, pos) + text.substring(pos + 1);
				myTextLabel.setText(text);
				myTextLabel.setDisplayedMnemonicIndex(pos);
			}
		}
		else {
			myTextLabel.setDisplayedMnemonicIndex(-1);
		}

		if (step.hasSubstep(value) && isSelectable) {
			myNextStepLabel.setVisible(true);
			final boolean isDark = ColorUtil.isDark(UIUtil.getListSelectionBackground());
			myNextStepLabel.setIcon(isSelected ? isDark ? NextStepInverted
					: NextStep
					: NextStepGrayed);
		}
		else {
			myNextStepLabel.setVisible(false);
			//myNextStepLabel.setIcon(PopupIcons.EMPTY_ICON);
		}

		if (isSelected) {
			setSelected(myNextStepLabel);
		}
		else {
			setDeselected(myNextStepLabel);
		}
	}


}

File: src/scratch/ide/popup/ScratchListElementRenderer.kt
package scratch.ide.popup

import com.intellij.openapi.ui.popup.ListItemDescriptor
import com.intellij.openapi.util.IconLoader
import com.intellij.ui.ColorUtil
import com.intellij.ui.popup.list.GroupedItemsListRenderer
import com.intellij.util.ui.UIUtil
import scratch.Scratch
import javax.swing.JList

fun createListItemDescriptor(myPopup: ScratchListPopup) = object: ListItemDescriptor<Scratch> {
    override fun getTextFor(value: Scratch) = myPopup.listStep.getTextFor(value)
    override fun getTooltipFor(value: Scratch): String? = null
    override fun getIconFor(value: Scratch) = myPopup.listStep.getIconFor(value)
    override fun hasSeparatorAboveOf(value: Scratch) = myPopup.listModel.isSeparatorAboveOf(value)
    override fun getCaptionAboveOf(value: Scratch) = myPopup.listModel.getCaptionAboveOf(value)
}

internal class ScratchListElementRenderer(
    private val myPopup: ScratchListPopup
): GroupedItemsListRenderer<Scratch>(createListItemDescriptor(myPopup)) {

    override fun customizeComponent(list: JList<out Scratch>, value: Scratch, isSelected: Boolean) {
        val step = myPopup.listStep
        val isSelectable = step.isSelectable(value)
        myTextLabel.isEnabled = isSelectable

        if (step.isMnemonicsNavigationEnabled) {
            val navigationFilter = step.mnemonicNavigationFilter
            val pos = navigationFilter?.getMnemonicPos(value) ?: -1
            if (pos != -1) {
                var text = myTextLabel.text
                text = text.substring(0, pos) + text.substring(pos + 1)
                myTextLabel.text = text
                myTextLabel.displayedMnemonicIndex = pos
            }
        } else {
            myTextLabel.displayedMnemonicIndex = -1
        }

        if (step.hasSubstep(value) && isSelectable) {
            myNextStepLabel.isVisible = true
            myNextStepLabel.icon = if (isSelected) {
                val isDark = ColorUtil.isDark(UIUtil.getListSelectionBackground(true))
                if (isDark) nextStepInverted else nextStep
            } else {
                nextStepGrayed
            }
        } else {
            myNextStepLabel.isVisible = false
            //myNextStepLabel.setIcon(PopupIcons.EMPTY_ICON);
        }

        if (isSelected) {
            setSelected(myNextStepLabel)
        } else {
            setDeselected(myNextStepLabel)
        }
    }

    companion object {
        val nextStep = IconLoader.getIcon("/icons/ide/nextStep.svg", IconLoader::class.java)
        val nextStepGrayed = IconLoader.getIcon("/icons/ide/nextStep_dark.png", IconLoader::class.java)
        val nextStepInverted = IconLoader.getIcon("/icons/ide/nextStepInverted.svg", IconLoader::class.java)
    }
}